From: Alexander Graf <agraf@suse.de>
Date: Sun, 07 Apr 2013 01:22:00 +0100
Subject: [PATCH] ARM: exynos: only use device tree when it exists
Patch-Mainline: No, will push later

On ARM, we can either boot systems using device tree, or using the legacy
machine ID based lookup. Some machines rely on the machine ID based
lookup to actually function while Linux is transitioning towards device tree
based machine configuration.

Unfortunately, during that process, some exynos5 code got added that
unconditionally accesses the device tree, breaking non device tree boards.

Fix it by only accessing the device tree when it's actually there.

Signed-off-by: Alexander Graf <agraf@suse.de>

Index: linux-3.9-rc5-master/arch/arm/mach-exynos/common.c
===================================================================
--- linux-3.9-rc5-master.orig/arch/arm/mach-exynos/common.c
+++ linux-3.9-rc5-master/arch/arm/mach-exynos/common.c
@@ -338,13 +338,16 @@ void __init exynos_init_io(struct map_de
 {
 	struct map_desc *iodesc = exynos_iodesc;
 	int iodesc_sz = ARRAY_SIZE(exynos_iodesc);
+
 #if defined(CONFIG_OF) && defined(CONFIG_ARCH_EXYNOS5)
-	unsigned long root = of_get_flat_dt_root();
+	if (initial_boot_params) {
+		unsigned long root = of_get_flat_dt_root();
 
-	/* initialize the io descriptors we need for initialization */
-	if (of_flat_dt_is_compatible(root, "samsung,exynos5440")) {
-		iodesc = exynos5440_iodesc;
-		iodesc_sz = ARRAY_SIZE(exynos5440_iodesc);
+		/* initialize the io descriptors we need for initialization */
+		if (of_flat_dt_is_compatible(root, "samsung,exynos5440")) {
+			iodesc = exynos5440_iodesc;
+			iodesc_sz = ARRAY_SIZE(exynos5440_iodesc);
+		}
 	}
 #endif
 
